blueslip: Make stacktrace more readable.

tommyip opened the pull request:
The stacktrace is now sourcemapped and each stackframe can be expanded to a code context.


**Testing Plan:** <!-- How have you tested? -->
Manual testing, unit tests

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![no-code-context](https://user-images.githubusercontent.com/18504232/59739985-45c99080-9299-11e9-82ed-d8481541b5fe.png)
![code-context](https://user-images.githubusercontent.com/18504232/59739995-4b26db00-9299-11e9-9dee-2fcc7d4e5890.png)



<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on static/js/click_handlers.js:
```diff
@@ -461,8 +461,8 @@ exports.initialize = function () {
     // this will hide the alerts that you click "x" on.
-    $("body").on("click", ".alert .exit", function () {
-        var $alert = $(this).closest(".alert");
+    $("body").on("click", ".alert-box>div .exit", function () {
```
We'd usually put spacing around the `>` here, but I'm also curious why this change is needed --- is there a new type of `.exit` widget?

tommyip replied:
The stacktrace popup doesn't use the `.alert` styling, so we need this for the exit button to work. (This is also why I extracted some bits out of `.alert` to `.alert-base` to share with stacktrace)

timabbott commented on static/styles/alerts.scss:
```diff
@@ -1,11 +1,32 @@
+    .faded {
+        opacity: 0.7;
+    }
+}
```
Can you split the extraction of `alert-base` into a prep refactor commit?  That'd make it easier to see the actual code changes.

timabbott commented on static/styles/alerts.scss:
```diff
@@ -66,17 +175,15 @@
-            color: hsl(0, 80%, 40%);
-            border: 1px solid hsl(0, 80%, 40%);
-            box-shadow: 0 0 2px hsl(0, 80%, 40%);
+            color: $alert-error-red;
```
Similarly, you should extract a prep commit for extracting these colors to variables.

timabbott commented on tools/webpack.config.ts:
```diff
@@ -184,6 +184,9 @@ export default (env?: string): webpack.Configuration => {
             },
+            headers: {
+                "Access-Control-Allow-Origin": "http://localhost:9991",
+            },
```
Can you explain why we need this change?  We might need to also have something related to `*.zulipdev.com`; otherwise realms you create with the "create realm" button won't work.

(Probably worth splitting to its own commit as well)

tommyip replied:
The network request made in https://github.com/zulip/zulip/pull/12608/files#diff-128e0920d73303bbd167e336de1ab806R169 is to `localhost:9994` but the script is running from `localhost:9991`, and different ports are considered different origins by the same-origin policy.

> We might need to also have something related to *.zulipdev.com

I'll have a look into that

tommyip replied:
Turns out `Access-Control-Allow-Origin` only allows the value `*`, `null` and an exact address, no regex or multiple addresses. What should I do here?

priyank-p replied:
We already have a proxy for webpack port; request made to `localhost:9991/webpack` goes to the webpack port. So just replacing  `localhost:9994` to `localhost:9991/webpack` should work? (Don't have time to check right now though)

tommyip replied:
The /webpack path only have the built bundles, not the file system. Or maybe I am missing something?

priyank-p replied:
Yeah, never mind you are right!

timabbott replied:
I think the right solution is we should add something in `zproject/dev_urls.py` to expose that part of the filesystem to the web, and then access that URL via port 9991?

tommyip replied:
@timabbott yeah that sounds good.

timabbott commented on static/js/blueslip.js:
```diff
@@ -96,20 +98,102 @@ exports.get_log = function blueslip_get_log() {
+function beautify_stackframe(stackframe) {
+    return stack_trace_gps
+        .getMappedLocation(stackframe)
+        .then(function (info) {
```
It might make sense to make this anonymous function extracted with a name, so that we can write some unit tests for its formatting logic.  It has a lot of tricky regexs/cases and seems like the kind of thing that's ideal for that sort of testing.

timabbott commented on static/js/blueslip.js:
```diff
@@ -96,20 +98,102 @@ exports.get_log = function blueslip_get_log() {
+
+    // We don't polyfill Promise but since this is available in debug
+    // mode only, assume the developer's browser is recent enough.
+    Promise.all(stacktrace).then(function (stackframes) {
```
Maybe expand this comment to explain why it's a promise?  (I assume it's because `beautify_stackframe` has to go over the network to get data?)

tommyip replied:
The stacktrace-gps library returns a promise in all of their APIs. I considered the other stacktrace sourcemapping libraries that doesn't use promises, but their browser support is not great.
* sourcemapped-stacktrace - chrome & firefox
* node-source-map-support - chrome only (primarily for node)

timabbott replied:
Sounds great; let's just expand the comment to explain that (including the reason it does so, which I'm pretty sure is because it might need to fetch the source map from the server)

timabbott commented:
Also, this looks awesome :)

tommyip commented:
Addressed Tim's review. Still have to work on dark mode.

timabbott commented:
I merged the first two refactor commits, thanks for splitting those out.

Casper is failing; my guess is it's because of ES6 syntax being used in non-transpiled code.  Depending on what the cause is, one possible solution is to migrate blueslip to typescript so that it is transpiled.

timabbott commented:
Oh, also, the Casper test failure might instead be caused by the `webpack` port issue we discussed above, since I think it uses zulipdev.com URLs.

tommyip commented:
Ready for review. The casper failures is properly something to do with ES6 usage though I don't really know where, casper is not exactly clear about that. Moving to typescript fixes things.

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,113 @@
+    // The stacktrace-gps library uses Promise in their APIs for fetching
+    // source maps from the server and the underlying mozilla/source-map
+    // library is compiled to wasm. We assume the developer's browser is recent
+    // enough to support both of these.
```
We’re supposed to work on IE11, which doesn’t provide `Promise` natively. But surely we already depend on twelve or fifteen libraries that pull in three different partially-compatible `Promise` polyfills? This is JS we’re talking about here.

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,113 @@
+    Promise.all(stacktrace).then(stackframes => {
+        const template = _.template($('.stacktrace #stackframe-template').html());
+        return stackframes.map(template).join('');
+    }).then(rendered_stackframes => {
```
The `.map(.then)` → `Promise.all` → `.then(.map)` → `.then` flow is unnecessarily awkward; there’s no reason to have so many different ways to apply a function to each thing. And if we’re going all-in on TypeScript, we should prefer `async` + `await` over manual `.then` applications.

```ts
export async function display_stacktrace(error: string, stack: string): Promise<void> {
    const ex = new Error();
    ex.stack = stack;
    const template = _.template($('.stacktrace #stackframe-template').html());
    const rendered_stackframes = (await Promise.all(
        ErrorStackParser.parse(ex).map(async stackframe =>
            template(clean_stackframe_data(
                await stack_trace_gps.getMappedLocation(stackframe)))))).join('');
    $('.stacktrace .stacktrace-content').html(rendered_stackframes);
    $('.stacktrace .message').html('<b>Exception</b> ' + error);
    $('.stacktrace').addClass('show');
}
```

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,113 @@
+    // library is compiled to wasm. We assume the developer's browser is recent
+    // enough to support both of these.
+    Promise.all(stacktrace).then(stackframes => {
+        const template = _.template($('.stacktrace #stackframe-template').html());
```
This is the first time we’d be using `_.template` in the main app. Why not use Handlebars like everything else?

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,113 @@
+        return stackframes.map(template).join('');
+    }).then(rendered_stackframes => {
+        $('.stacktrace .stacktrace-content').html(rendered_stackframes);
+        $('.stacktrace .message').html('<b>Exception</b> ' + error);
```
Is `error` escaped for HTML?

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,113 @@
+            const rendered_context = lines.slice(lo_line_num, hi_line_num)
+                .map((line: string) => {
+                    line = line.replace(/ /g, '&nbsp;');
+                    line = line.replace(/\t/g, '&nbsp;&nbsp;&nbsp;&nbsp;');
```
`&nbsp;` replacements are annoying for copy+paste purposes. Can we use `<pre>` or CSS `white-space: pre`/`pre-wrap`?

tommyip replied:
Yeah. Though copying will still be a bit annoying due to the line numbers.

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,113 @@
+    const fn_name = full_fn_name.split('.');
+
+    fn_name[fn_name.length - 1] = '<span class="function-name">'
+        + fn_name[fn_name.length - 1].replace('<', '&lt;').replace('>', '&gt;')
```
This seems like a weird thing to do: we’re HTML-escaping only the last element of the array, and only partially at that (no `&` ↦ `&amp;`)?

tommyip replied:
This is only for the case `abc.<anonymous>`, otherwise the names are always just alphanumeric and underscores.

andersk replied:
I’d be more comfortable with an implementation that’s guaranteed to escape everything than an implementation that hasn’t yet been found to forget to escape anything important.

It would be even better to let the template take care of escaping: split at the last `.` and pass the two parts to the template separately.

tommyip replied:
Done

tommyip commented:
Addressed @andersk review.

timabbott commented:
Can you rebase this?  I just merged your big frontend deps upgrade, so probably some conflicts.

tommyip commented:
Rebased, also fixed a mistake.
```diff
diff --git a/static/js/tsconfig.json b/static/js/tsconfig.json
index befa07654..0eecc0209 100644
--- a/static/js/tsconfig.json
+++ b/static/js/tsconfig.json
@@ -31,6 +31,6 @@
         "noImplicitReturns": true,
         "noFallthroughCasesInSwitch": true,
 
-        "typeRoots": ["node_modules/@types", "../../static/js/js_typings"]
+        "typeRoots": ["../../node_modules/@types", "js_typings"]
     }
 }
```

timabbott commented:
How does `error-stack-parser` get bundled in this implementation?  

Can you also comment on how you resolved Anders' comments?  I'm not sure

andersk commented:
> How does `error-stack-parser` get bundled in this implementation?

Webpack bundles everything that you transitively `import`/`require()` from anywhere. (The pattern where we import every library into the global namespace from `static/js/bundles` doesn’t actually make sense in a Webpack world, except as a transitional device from the pre-Webpack world. We should take every convenient opportunity to move away from it.)

tommyip commented:
> Can you also comment on how you resolved Anders' comments? I'm not sure how you're addressing the Promises/IE11 question, for example.

```js
// The stacktrace-gps library uses Promise in their APIs for fetching
// source maps from the server and the underlying mozilla/source-map
// library is compiled to wasm. We assume the developer's browser is recent
// enough to support both of these.
```
The stacktrace popup is only available in debug mode, so I think assuming a browser that support both promise and wasm is reasonable?

timabbott commented:
> The stacktrace popup is only available in debug mode, so I think assuming a browser that support both promise and wasm is reasonable? I could perhaps add a warning/fallback if they are not available.

We also need to be careful -- the `blueslip` module is imported everywhere, so if it has syntax that doesn't exist in prod, it'll break there, even if we're not calling the code.

tommyip commented:
We can use something like https://github.com/nippur72/ifdef-loader to conditionally remove the debugging code during compilation. Not sure whether this is needed though since this works fine in casper and that doesn't support any ES6.

timabbott commented:
Is it difficult to just add a promise polyfill so that we don't need to worry about it?  As Anders mentioned, we probably already have one in the bundle anyway, so the cost is likely ~0.

timabbott commented:
(But if you want to test with a browser that doesn't support Promise triggering a JS exception on a portico page, that'd also work)

tommyip commented:
Added the polyfill.

andersk commented on yarn.lock:
```diff
@@ -2196,6 +2208,11 @@ copy-descriptor@^0.1.0:
+  version "3.1.4"
+  resolved "https://registry.yarnpkg.com/core-js/-/core-js-3.1.4.tgz#3a2837fc48e582e1ae25907afcd6cf03b0cc7a07"
+  integrity sha512-YNZN8lt82XIMLnLirj9MhKDFZHalwzzrL9YLt6eb0T5D0EDl4IQ90IGkua8mHbnxNrkj1d8hbdizMc0Qmg1WnQ==
+
```
SimpleBar already pulls in core-js 3.0.1 as `^3.0.1`. Can we convince it to use 3.1.4 too?

(And I’d prefer `core-js@^3.1.4` over `core-js@3.1.4` so as to avoid reintroducing a similar issue later.)

tommyip replied:
I think we can do something like https://github.com/zulip/zulip/blob/master/package.json#L87 again.

timabbott commented:
@tommyip sorry for the merge conflicts, but I just merged a bunch of #12676 which you'll need to rebase with respect to.  Shouldn't be a big effort.

tommyip commented:
The CircleCI failure is unrelated to this PR. My handlebars-loader PR missed one of the template path that needed changing, fixed in #12706. Though it is interesting this error is only showing up here.

tommyip commented:
Test passed.

timabbott commented:
Rebased and pushed back here; @andersk can you do a last check on the toolchain changes here?

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,102 @@
+// we assume the developer's browser is recent enough to support it.
+const stack_trace_gps = new StackTraceGPS();
+
+export async function display_stacktrace(error: string, stack: string): Promise<void> {
```
Now that this goes through Babel, we need to `yarn add regenerator-runtime` for `async`.

andersk commented on zproject/dev_urls.py:
```diff
@@ -14,6 +14,10 @@
 urls = [
+    # Serve node_modules for code context in blueslip's stacktrace
+    url(r'^node_modules/(?P<path>.*)$',
+        serve, {'document_root': os.path.join(settings.DEPLOY_ROOT, 'node_modules')}),
```
Hold on. Doesn’t the source map already contain all the code? Why do we need to fetch it separately?

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,102 @@
+        })
+    );
+
+    const rendered = templates.render('blueslip_stacktrace', { stackframes });
```
This needs an update for the Handlebars `require` migration #12761.

andersk commented:
I noticed that there’s no way to scroll the blueslip error overlay if it gets taller than the screen height. We probably want an `overflow: auto` container somewhere.

tommyip commented:
Updated. Waiting for zstub from #12858 so I can stub out the wasm require call, otherwise the node test won't run.

andersk commented on static/js/tsconfig.json:
```diff
@@ -13,6 +13,11 @@
+        // Include the definitions (not polyfill)
+        "lib": ["es2015", "dom"],
+        // Allow tsc to find types included with packages and not in @types/*
+        "moduleResolution": "node",
```
Oh _that’s_ what this is for. We can `yarn remove @types/source-map` then.

tommyip replied:
When did we install `@types/source-map`?

andersk replied:
#12838. With neither `@types/source-map` nor `--moduleResolution=node`, `tsc` wouldn’t run.

tommyip replied:
Oh cool.

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,122 @@
+                    lo_line_num += 1;
+                    return '<div class="' + (lo_line_num === line_number ? 'focus-line' : '') + '">'
+                        + '<span class="line-number">' + lo_line_num + '</span>&nbsp;'
+                        + line + '</div>';
```
As discussed before, do not try to generate HTML with string concatenation. This doesn’t escape anything, so it’s wrong. It may or may not be a security vulnerability depending on whether an attacker can find a way to control the input, but it’s wrong either way.

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,122 @@
+                    // We don't know the exact file path due to the node_modules
+                    // hash so we have to find it.
+                    for (const source of consumer.sources) {
+                        if (source.match(full_path + '$')) {
```
Smashing a dollar sign onto a random string doesn’t make it a regular expression.

andersk commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,122 @@
+
+// The stacktrace-gps library relies on the mozilla/source-map WASM library,
+// we assume the developer's browser is recent enough to support it.
+const stack_trace_gps = new StackTraceGPS();
```
Now I’m curious…how does this library avoid doing the `.wasm` URL dance we just had to do?

tommyip replied:
They are using an old version of source-map which doesn't use the wasm mappings.

tommyip commented on static/js/blueslip_stacktrace.ts:
```diff
@@ -0,0 +1,122 @@
+import StackTraceGPS, { StackFrame } from 'stacktrace-gps';
+import ErrorStackParser from 'error-stack-parser';
+const render_blueslip_stacktrace = require('../templates/blueslip_stacktrace.hbs');
+const source_map_mappings = require('source-map/lib/mappings.wasm');
```
Have to add a declaration
```ts
declare module "*.hbs" {
    const render: (args: { [key: string]: any }) => string;
    export = render;
}
```
otherwise typescript won't recognize it.

tommyip commented:
Fixed.

andersk commented:
It seems like when multiple errors are thrown, new errors hide the old ones instead of being displayed below them like in your original screenshot?

andersk commented:
Rebased to address my minor comments and greatly simplify some of the logic. Still needs some work to address this, though:

> It seems like when multiple errors are thrown, new errors hide the old ones instead of being displayed below them like in your original screenshot?

andersk commented:
Restructured to allow displaying multiple errors again. I think this is ready for another look from @timabbott.

timabbott commented:
Looks great, merged, thanks @tommyip and @andersk!  Sorry this took so long to get integrated.

timabbott commented:
I'm really looking forward to being able to use it though :)
